File: examples/FSharp/OpenGL/Tutorial 1.3 - Abstractions/VertexArrayObject.fs
namespace Tutorial

open Silk.NET.OpenGL

type VertexArrayObject = {
    gl: GL
    handle: uint32
}

module VertexArrayObject =
    let bind (vao: VertexArrayObject) =
        vao.gl.BindVertexArray(vao.handle)


    let create (gl: GL) vbo ebo =
        let handle = gl.GenVertexArray()
        let vao = {
            gl = gl
            handle = handle
        }

        bind vao
        BufferObject.bind vbo
        BufferObject.bind ebo
        vao


    let attributePointer index count (dataType: GLEnum) vertexSize offset vao =
        let typeSize =
            match dataType with
            | GLEnum.Float -> uint32 sizeof<float32>
            | GLEnum.UnsignedInt -> uint32 sizeof<uint32>
            | _ -> failwithf "attributePointer cannot set data of type: %A" dataType

        vao.gl.VertexAttribPointer(index, count, dataType, false, vertexSize * typeSize, offset * int typeSize)
        vao.gl.EnableVertexAttribArray(index)

    let destroy (vao: VertexArrayObject) =
        vao.gl.DeleteVertexArray(vao.handle)

File: examples/CSharp/OpenGL Tutorials/Tutorial 4.1 - Model Loading/VertexArrayObject.cs
using Silk.NET.OpenGL;
using System;

namespace Tutorial
{
    public class VertexArrayObject<TVertexType, TIndexType> : IDisposable
        where TVertexType : unmanaged
        where TIndexType : unmanaged
    {
        private uint _handle;
        private GL _gl;

        public VertexArrayObject(GL gl, BufferObject<TVertexType> vbo, BufferObject<TIndexType> ebo)
        {
            _gl = gl;

            _handle = _gl.GenVertexArray();
            Bind();
            vbo.Bind();
            ebo.Bind();
        }

        public unsafe void VertexAttributePointer(uint index, int count, VertexAttribPointerType type, uint vertexSize, int offSet)
        {
            _gl.VertexAttribPointer(index, count, type, false, vertexSize * (uint) sizeof(TVertexType), (void*) (offSet * sizeof(TVertexType)));
            _gl.EnableVertexAttribArray(index);
        }

        public void Bind()
        {
            _gl.BindVertexArray(_handle);
        }

        public void Dispose()
        {
            _gl.DeleteVertexArray(_handle);
        }
    }
}
